presto-main/src/main/java/com/facebook/presto/cost/SemiJoinStatsCalculator.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cost;
+
+import com.facebook.presto.sql.planner.Symbol;
+
+import static com.facebook.presto.cost.PlanNodeStatsEstimate.UNKNOWN_STATS;
+import static java.lang.Double.NaN;
+
+public class SemiJoinStatsCalculator
+{
+    // TODO implementation does not take into account overlapping of ranges for source and filtering source.
+    //      Basically it works as low and high values were the same for source and filteringSource and just looks at NDVs.
+
+    public PlanNodeStatsEstimate computeSemiJoin(PlanNodeStatsEstimate sourceStats, PlanNodeStatsEstimate filteringSourceStats, Symbol sourceJoinSymbol, Symbol filteringSourceJoinSymbol)
+    {
+        SymbolStatsEstimate sourceJoinSymbolStats = sourceStats.getSymbolStatistics(sourceJoinSymbol);
+        SymbolStatsEstimate filteringSourceJoinSymbolStats = filteringSourceStats.getSymbolStatistics(filteringSourceJoinSymbol);
+
+        // TODO: use range methods when they have defined (and consistent) semantics
+        double sourceNDV = sourceJoinSymbolStats.getDistinctValuesCount();
+        double filteringSourceNDV = filteringSourceJoinSymbolStats.getDistinctValuesCount();
+
+        PlanNodeStatsEstimate outputStats = sourceStats;
+        if (sourceNDV > filteringSourceNDV) {
+            double scaleFactor = (1 - sourceJoinSymbolStats.getNullsFraction()) * (filteringSourceNDV / sourceNDV);
+            outputStats = outputStats.mapSymbolColumnStatistics(sourceJoinSymbol, columnStats ->
+                    SymbolStatsEstimate.buildFrom(columnStats)
+                            .setLowValue(sourceJoinSymbolStats.getLowValue())
```

Comment:
Isn't it no-op? `columnStats` is the same as `sourceJoinSymbolStats`.
In fact, I find it more readable if you avoid `mapSymbolColumnStatistics`:
```
newStats = SymbolStatsEstimate.buildFrom(sourceJoinSymbolStats)
    .setNullsFraction(0)
    .setDistinctValuesCount(filteringSourceNDV)
    .build();
outputStats = outputStats.mapSymbolColumnStatistics(sourceJoinSymbol, oldStats -> newStats);
```